### PR TITLE
feat(internal): fixedRegistries

### PR DESCRIPTION
**Pull request by rarkins**
## Changes:

- Adds "fixed" registryUrls for multiple datasources
- Add explicit "disallowed" registryUrls so that attempts to override them can be warned about

## Context:

- Provide warning when `registryUrls` are used mistakenly
- Provide more transparency about fixed registryUrls
- Part of the steps necessary for datasource proxying and cache

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added unit tests, or
- [ ] No new tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/datasource/cdnjs/index.ts [hunk 1]
```diff
@@ -24,10 +26,11 @@ interface CdnjsResponse {
 
 export async function getReleases({
   lookupName,
+  registryUrl,
 }: GetReleasesConfig): Promise<ReleaseResult | null> {
   // Each library contains multiple assets, so we cache at the library level instead of per-asset
   const library = lookupName.split('/')[0];
-  const url = `https://api.cdnjs.com/libraries/${library}?fields=homepage,repository,assets`;
+  const url = `${registryUrl}libraries/${library}?fields=homepage,repository,assets`;
```
Comment (rarkins) [inline lib/datasource/cdnjs/index.ts:33]: Should we use a url function for this resolving?

File: lib/datasource/npm/index.ts [hunk 2]
```diff
@@ -5,3 +5,4 @@ export { getReleases } from './releases';
 export { getNpmrc, setNpmrc } from './npmrc';
 export { id } from './common';
 export const defaultVersioning = npmVersioning.id;
+export const registryUrlRestriction = 'disallowed';
```
Comment (rarkins) [inline lib/datasource/npm/index.ts:8]: This needs refactoring later to use `defaultRegistryUrls` and `rangeStrategy=first`. But first we need to stop using `.npmrc` directly

File: lib/datasource/types.ts [hunk 3]
```diff
@@ -63,6 +63,17 @@ export interface DatasourceApi {
   defaultRegistryUrls?: string[];
   defaultVersioning?: string;
   defaultConfig?: Record<string, unknown>;
+
+  // registryStrategy=first means only the first registryUrl will be tried and others ignored
+  // registryStrategy=hunt means registryUrls will be tried in order until one returns a result
+  // registryStrategy=merge means all registryUrls will be tried and the results merged if more than one returns a result
```
Comment (viceice) [inline lib/datasource/types.ts:69]: convert to jsdoc style

File: lib/datasource/types.ts [hunk 4]
```diff
@@ -63,6 +63,17 @@ export interface DatasourceApi {
...
   registryStrategy?: 'first' | 'hunt' | 'merge';
+
+  // registryUrlRestriction=fixed means the default registryUrl settings can't be overridden
+  // registryUrlRestriction=disallowed means that registryUrls are not applicable to this datasource
+  // If registryUrlRestriction is unspecified, it means custom registryUrls are allowed (no retriction)
```
Comment (viceice) [inline lib/datasource/types.ts:74]: jsdoc style

File: lib/datasource/types.ts [hunk 5]
```diff
@@ -63,6 +63,17 @@ export interface DatasourceApi {
...
+  registryUrlRestriction?: 'fixed' | 'disallowed';
+
+  // caching=true means caching will be performed by the datasource index instead of the datasource implementation
```
Comment (viceice) [inline lib/datasource/types.ts:77]: this too

File: lib/datasource/npm/index.ts (hunk 2 above)
Comment (viceice) [reply lib/datasource/npm/index.ts:8]: Doesn't this now blocks custom npm registries? I'm overriding this on many internal projects.

File: lib/datasource/galaxy/index.spec.ts [hunk 6]
```diff
@@ -89,6 +89,18 @@ describe('datasource/galaxy', () => {
...
+      httpMock
+        .scope(baseUrl)
+        .get('/api/v1/roles/?owner__username=yatesr2&name=timezone')
+        .reply(200, res1);
+      const res = await getPkgReleases({
+        datasource,
+        depName: 'yatesr2.timezone',
+        registryUrls: ['https://google.com/'],
+      });
+      expect(res).not.toBeNull();
```
Comment (viceice) [inline lib/datasource/galaxy/index.spec.ts:102]: We should add `logger.warn` call snapshot test to verify

File: lib/datasource/cdnjs/index.ts (hunk 1 above)
Comment (viceice) [reply lib/datasource/cdnjs/index.ts:33]: yes, but let's do this in separate pr

File: lib/datasource/npm/index.ts (hunk 2 above)
Comment (rarkins) [reply lib/datasource/npm/index.ts:8]: Wasn't intentional and can be reverted.

Comment (rarkins) [reply lib/datasource/npm/index.ts:8]: I checked and there's no `registryUrls` use directly in lib/datasource/npm. What approach do you use to override it?

Comment (viceice) [reply lib/datasource/npm/index.ts]: Ok, so how get the registry url passed from npmrc of repo?

Comment (rarkins) [reply lib/datasource/npm/index.ts]: There's an npm lib that actually reads from disk as part of the npm datasource

Comment (viceice) [reply lib/datasource/npm/index.ts]: So maybe we should move that logic to manager in a future pr.

**Review by viceice — APPROVED** [review]
Make some dry runs to be sure it works as expected?

**Review by zharinov — APPROVED** [review]
LGTM

**Pull request merged**